src/test/ensemble_flow_custom_node_tests.cpp

```diff
@@ -4264,3 +4432,146 @@ TEST_F(EnsembleFlowCustomNodePipelineExecutionTest, DemultiplexerConnectedToNhwc
+        return 0;
+    }
+    static int release(void* ptr, void* customNodeLibraryInternalManager) {
+        free(ptr);
+        return 0;
+    }
+};
+
+TEST_F(EnsembleFlowCustomNodePipelineExecutionTest, MultipleDeinitializeCallsOnRetire) {
+    // Nodes
+    // request   custom    response
+    //  O--------->O---------->O
+    //          add-sub
+    ConstructorEnabledModelManager manager;
+    PipelineFactory factory;
+
+    const std::vector<float> inputValues{7.8, -2.4, 1.9, 8.7, -2.4, 3.5};
+    this->prepareRequest(inputValues);
+
+    const float addValue = 0.9;
+    const float subValue = 7.3;
+
+    auto mockedLibrary = createLibraryMock<LibraryCountDeinitialize>();
+    mockedLibrary.execute = library.execute;
+    mockedLibrary.getInputsInfo = library.getInputsInfo;
+    mockedLibrary.getOutputsInfo = library.getOutputsInfo;
+    mockedLibrary.release = library.release;
+
+    LibraryCountDeinitialize::deinitializeCounter = 0;
+
+    std::vector<NodeInfo> info{
+        {NodeKind::ENTRY, ENTRY_NODE_NAME, "", std::nullopt, {{pipelineInputName, pipelineInputName}}},
+        {NodeKind::CUSTOM, "custom_node", "", std::nullopt, {{customNodeOutputName, customNodeOutputName}},
+            std::nullopt, {}, mockedLibrary, parameters_t{{"add_value", std::to_string(addValue)}, {"sub_value", std::to_string(subValue)}}},
+        {NodeKind::CUSTOM, "custom_node_2", "", std::nullopt, {{customNodeOutputName, customNodeOutputName}},
+            std::nullopt, {}, mockedLibrary, parameters_t{{"add_value", std::to_string(addValue)}, {"sub_value", std::to_string(subValue)}}},
+        {NodeKind::CUSTOM, "custom_node_3", "", std::nullopt, {{customNodeOutputName, customNodeOutputName}},
+            std::nullopt, {}, mockedLibrary, parameters_t{{"add_value", std::to_string(addValue)}, {"sub_value", std::to_string(subValue)}}},
+        {NodeKind::EXIT, EXIT_NODE_NAME},
+    };
+
+    pipeline_connections_t connections;
+
+    // request (pipelineInputName) O--------->O custom node (customNodeInputName)
+    connections["custom_node"] = {
+        {ENTRY_NODE_NAME, {{pipelineInputName, customNodeInputName}}}};
+
+    // custom node (customNodeOutputName) O--------->O custom node 2 (customNodeInputName)
+    connections["custom_node_2"] = {
+        {"custom_node", {{customNodeOutputName, customNodeInputName}}}};
+
+    // custom node 2 (customNodeOutputName) O--------->O custom node 3 (customNodeInputName)
+    connections["custom_node_3"] = {
+        {"custom_node_2", {{customNodeOutputName, customNodeInputName}}}};
+
+    // custom node (customNodeOutputName) O--------->O response (pipelineOutputName)
+    connections[EXIT_NODE_NAME] = {
+        {"custom_node_3", {{customNodeOutputName, pipelineOutputName}}}};
+
+    ASSERT_EQ(factory.createDefinition("my_new_pipeline", info, connections, manager), StatusCode::OK);
+
+    factory.retireOtherThan({}, manager);
+    ASSERT_EQ(LibraryCountDeinitialize::deinitializeCounter, 3);
```

Comment:
While this test is correct as effectively the underlying library just have global deinitializeCounter, I wonder if it might be confusing to someone trying to grasp this concept by studying our test. So maybe at the very least add comment here that while there is global counter each node has effectively 1 internalManagerInitialized?

Its clear to me what is happening here but it might be not for others.